[docs] Manifest preload docs (#1031)
earth-mover/icechunk: docs/docs/icechunk-python/performance.md

```python exec="on" session="perf" source="material-block"
print(ic.Repository.open(storage).config.manifest)
```

## Preloading manifests

> While [manifest splitting](./performance.md#splitting-manifests) is a great way to control the size of manifests, it can be useful to configure the manner in which manifests are loaded. In Icechunk manifests are loaded lazily by default, meaning that when you read a chunk, Icechunk will only load the manifest for that chunk when it is needed to fetch the chunk data. While this is good for memory performance, it can be a bottleneck for read performance when you are reading a lot of chunks.

Reviewer: ```suggestion
While [manifest splitting](./performance.md#splitting-manifests) is a great way to control the size of manifests, it can be useful to configure the manner in which manifests are loaded. In Icechunk manifests are loaded lazily by default, meaning that when you read a chunk, Icechunk will only load the manifest for that chunk when it is needed to fetch the chunk data. While this is good for memory performance, it can increase the latency to first the first elements of an array. Once a manifest has been loaded, it will usually remain in memory for future chunks in the same manifest.
```

To address this, Icechunk provides a way to preload manifests at the time of opening a Session, loading all manifests that match specific conditions into the cache. This means when it is time to read a chunk, the manifest will already be in the cache and the chunk can be found without having to first load the manifest from storage.

### Configuring preloading

To configure manifest preloading, you can use the `ManifestPreloadConfig` class to specify the `ic.RepositoryConfig.manifest.preload` field.

```python exec="on" session="perf" source="material-block"
preload_config = ic.ManifestPreloadConfig(

>     preload_if=ic.ManifestPreloadCondition.name_matches("x"),  # preload all manifests that match the regex "x"

Reviewer: I think name_matches takes a regex, so this may match much more than they want? I don't remember if we do full match or we match anywhere in the name. So, this may need something like `("^x$")`. But I'd have to check how we match it, maybe you already did. It would be nice to add it to this docs.

Author: You are totally right, I'll refine this

Author: done

Preloading can also be limited to manifests that are within a limited size range. This can be useful to limit the amount of memory used by the preload cache, when some manifest may be very large. This can be configured using the `ic.RepositoryConfig.manifest.preload.num_refs` field.

```python exec="on" session="perf" source="material-block"
preload_config = ic.ManifestPreloadConfig(
    preload_if=ic.ManifestPreloadCondition.and_conditions(
        ic.ManifestPreloadCondition.name_matches("x"),
        ic.ManifestPreloadCondition.num_refs(1000, 10000),
    )
)
```

This will preload all manifests that match the array name "x" and have between 1000 and 10000 chunk references.

Lastly, the number of total manifests that can be preloaded can be limited using the `ic.RepositoryConfig.manifest.preload.max_total_refs` field.

```python exec="on" session="perf" source="material-block"
preload_config = ic.ManifestPreloadConfig(
    preload_if=ic.ManifestPreloadCondition.name_matches("x"),
    max_total_refs=10000,
)
```

This will preload all manifests that match the array name "x" while the number of total chunk references that have been preloaded is less than 10000.

!!! important

>     Once you find a preload configuration you like, remember to persist it on-disk using `repo.save_config`. It will still be able to be overridden at runtime for different applications.

Reviewer 2: ```suggestion
    Once you find a preload configuration you like, remember to persist it on-disk using `repo.save_config`. The saved config can be overridden at runtime for different applications.
```
